Serialization Refactor

antonydellavecchia opened the pull request:
Refactor serialization to use a new function called `type_params` which helps build a stack of type dependencies which will allow for serializing type information before the data.

### Release notes:
manually copy from https://github.com/oscar-system/Oscar.jl/pull/4162#issuecomment-2893356751

antonydellavecchia committed:
some progress on simplifying ref handling in IPC Serialization

antonydellavecchia committed:
some progress

antonydellavecchia committed:
moving things around

antonydellavecchia committed:
type params function

antonydellavecchia committed:
some progress on refactoring

antonydellavecchia committed:
some progress

antonydellavecchia committed:
more progress

antonydellavecchia committed:
almost completed uni variate polynomials

antonydellavecchia committed:
uni vairate polynomials serialization passing

antonydellavecchia committed:
MPolyRing tests passing

antonydellavecchia committed:
moving things around

antonydellavecchia committed:
grading needs containers

antonydellavecchia committed:
not sure where i left this

antonydellavecchia committed:
new load_type_params

antonydellavecchia committed:
bug with julia 1.11.1

antonydellavecchia committed:
vectors + polynomials working

antonydellavecchia committed:
some progress on tuples

antonydellavecchia committed:
reworking save_type_params

antonydellavecchia committed:
tidying

antonydellavecchia committed:
dicts working

antonydellavecchia committed:
containers completed, need some upgrades though

antonydellavecchia committed:
tidying

antonydellavecchia committed:
broke everything

antonydellavecchia committed:
fixed containers and polynomials

antonydellavecchia committed:
some issues with field embeddings

antonydellavecchia committed:
fields tests working now

antonydellavecchia committed:
algebras working

antonydellavecchia committed:
started on groups

lgoettgens commented:
This PR seems to have diverged a lot from current master due to other changes to the serialization code in the meantime. Could you try to resolve the conflicts / rebase this onto a more current master, so that one can more easily follow what's happening and CI runs again?

I am a bit afraid that this just continues to diverge even further in the future, which makes it harder to resolve conflicts once you want to merge it.

antonydellavecchia committed:
Merge branch 'master' into adv/ipc-serialize

lgoettgens commented on src/Combinatorics/SimplicialComplexes.jl:
```diff
@@ -670,7 +670,7 @@ function automorphism_group(K::SimplicialComplex; action=:on_vertices)
 
 @doc raw"""
-    on_simplicial_complex(K::SimplicialComplex, g::PermGroupElem)
+    on_simplicial_complex(K::SimplicialComplex, g::PermGroupElem; action=:on_vertices)
```
is this maybe a conflict resolution artifact? it does not match the signature of the function this docstring is attached to

antonydellavecchia committed:
fix, inconsistencies from merge

antonydellavecchia committed:
crawling through groups

antonydellavecchia commented on src/Serialization/GAP.jl:
```diff
@@ -125,43 +143,47 @@ install_GAP_deserialization(
+    return G
+  end)
+
+install_GAP_deserialization(
```
@ThomasBreuer  I think I need your help here, and it also makes sense for you to understand what I am doing here anyway.

I am currently try to get the serialization refactor working for gap objects, and the first one that causes an issue is when we try and save subgroups of free groups


If you see above I have a new function called `type_params` and this handles when params need to be added to the type description for example saving a subgroup of a free group currently looks like this

```
{
  "_ns": {
    "Oscar": [
      "https://github.com/oscar-system/Oscar.jl",
      "1.3.0-DEV-0324e87a589a2c2c5b6f4b0db4634b14477250ae"
    ]
  },
  "_refs": {
    "fef8f069-d2db-4c88-9118-f2b7ba1c3999": {
      "_type": "GapObj",
      "data": {
        "GapType": "IsFreeGroup",
        "names": [
          "f1",
          "f2"
        ],
        "wfilt": "IsSyllableWordsFamily"
      }
    }
  },
  "_type": {
    "name": "GapObj",
    "params": "fef8f069-d2db-4c88-9118-f2b7ba1c3999"
  },
  "data": {
    "GapType": "IsFreeGroup",
    "gens": [
      [
        "1",
        "1"
      ]
    ]
  },
  "id": "ab56be5f-1bf6-4fec-a3d5-35cb59b6096b"
}

```

Note how there is no longer a freeGroup key in the data section and that the type information of the GapObj now has a params key.

The issue I am having now is that I need to pass the free group down, but this means I need two different function signatures. 

But I am guessing these cannot share the same namespace in GAP?

please let me know if you have any questions about some of the changes here

ThomasBreuer replied:
@antonydellavecchia I am not sure whether I understand the aim of the proposed changes.
A subgroup of a full free group is best described in terms of generators relative to the full free group, hence the "old" approach first serializes the full free group and then the generators of the subgroup in question, and the deserialization first reconstructs the full free group and then the subgroup generators w.r.t. the full group.

The idea of the "new" approach seems to be that the function `type_params` shall describe the object(s) to be serialized together with the data of the given object.
The problem is that `type_params` for a `GapObj` has to deal with a lot of different `GapObj`s.
In the "old" serialization and deserialization of GAP objects, the approach using `install_GAP_serialization` and `install_GAP_deserialization` yields small Julia functions for different kinds of GAP objects. One could think of an analogous `install_GAP_type_params` in order to achieve this for `type_params` iin the "new" approach, but what is the advantage of using `type_params`, compared to the "old" (recursive) serialization approach?

Just for curiosity: How does the "new" approach handle objects that need more than one dependent object to be serialized?
(For example, the most natural way to serialize a map seems to be to serialize domain and codomain and information that defines the function in question.)
These dependent objects are stored in `_refs`, their keys in `_refs` shall be referenced somewhere in `_type`, but how does the deserialization code know which is which? In the "old" approach, the references to the keys in `_refs` are stored in fields inside `data` (like the `freeGroup` field), whose meanings are defined by the serialization.

antonydellavecchia replied:
The idea is that type params will do a shallow pass through the object and obtain all necessary type information for the object. This has many benefits, the main one being that this would allow us to put all type information at the top of the file and hence we would be able to read in all params for the object before we get to the data in a linear way. 

here is another example of how a the new format looks for a MPolyAnyMap.
You can see that the params section can also have keys, i.e. we can have params that are dictionnaries if necessary
(this does not yet include the linear ordering)

```
{
  "_ns": {
    "Oscar": [
      "https://github.com/oscar-system/Oscar.jl",
      "1.3.0-DEV-98fc56d6f49a6eacecd525b01c0e543857608f58"
    ]
  },
  "_refs": {
    "bede6628-38bf-4803-88df-2ed7487fc8c8": {
      "_type": {
        "name": "MPolyRing",
        "params": {
          "_type": "QQField"
        }
      },
      "data": {
        "symbols": [
          "s",
          "t",
          "m"
        ]
      }
    },
    "fc49a9a8-91af-4b54-888b-58b4f2dd1958": {
      "_type": {
        "name": "MPolyRing",
        "params": {
          "_type": "QQField"
        }
      },
      "data": {
        "symbols": [
          "x",
          "y"
        ]
      }
    }
  },
  "_type": {
    "name": "MPolyAnyMap",
    "params": {
      "codomain": {
        "name": "MPolyRing",
        "params": "bede6628-38bf-4803-88df-2ed7487fc8c8"
      },
      "domain": {
        "name": "MPolyRing",
        "params": "fc49a9a8-91af-4b54-888b-58b4f2dd1958"
      }
    }
  },
  "data": {
    "images": [
      [
        [
          [
            "1",
            "1",
            "0"
          ],
          "1"
        ]
      ],
      [
        [
          [
            "0",
            "0",
            "1"
          ],
          "1"
        ]
      ]
    ]
  }
}
```

ThomasBreuer replied:
Thanks @antonydellavecchia .
Thus the rule is that it is recommended (or mandatory?) to "announce" the dependent objects via `type_params`.

Then the interface for the (de)serialization for a given object type `T` (here: `GapObj`) consists of
- a `save_object` method with second argument `T`,
- a `load_object` method with two or three arguments, the second argument being `T`,
- a `type_params` method with one argument of type `T`, returning `S, X` where `X` is either `nothing` or an object of type `S` (a dictionary or a `GapObj`); in the latter case, there must be a three argument method for `load_object` that takes `X` as its third argument.

If this is correct then I propose a function `install_GAP_type_params` as sketched above for the case of `GapObj`. I can provide the code for that.

antonydellavecchia replied:
The rule will become mandatory.

the load object third parameters will exactly be the second entry of the output from `type_params`

the output of `type_params` should be a tuple where the first entry is the type and the second entry are the params for that type, for example above type_params with return the type  
```
PolyAnyMap, Dict(:codomain => QQ[:s, :t, :m], :domain => QQ[:x, :y])
```

for a  group element g It would be some thing like 

```
GroupElem, parent(g)
```

and yes exactly, if the second output of `type_params` is nothing, then that particular load_object method only needs two parameters

antonydellavecchia committed:
matrices working

antonydellavecchia committed:
MPolyMap Working

antonydellavecchia committed:
some progress on polyhedral objects

ThomasBreuer committed:
adjust GAP serialization to the refactoring

- introduce `install_GAP_type_params` that implements
  `type_params` methods based on GAP filters
- adjust code for `IsFreeGroup` and `IsSubgroupFpGroup`;
  now their tests work again

ThomasBreuer commented:
The tests for pc groups are still disabled. Here the problem is that `load_object(s, Vector, (Tuple, [Int, Int, [Vector, Int]]), :comm_rels)` does not work anymore.

antonydellavecchia committed:
GAP object serialization working

antonydellavecchia commented:
> The tests for pc groups are still disabled. Here the problem is that `load_object(s, Vector, (Tuple, [Int, Int, [Vector, Int]]), :comm_rels)` does not work anymore.

Thanks for that. I've made the loading of vectors a little more intuitive. 
Please see the changes I've just pushed to get the GAP object serialization working

ThomasBreuer commented:
> Please see the changes I've just pushed to get the GAP object serialization working

Thanks @antonydellavecchia.
~~I am going to adjust the serialization for GAP pc groups.~~
Now the tests for GAP pc groups pass.

antonydellavecchia committed:
groups working

antonydellavecchia committed:
removed false kwarg in docs

antonydellavecchia committed:
Lie Theory serialization working

lgoettgens commented on src/Serialization/LieTheory.jl:
```diff
@@ -112,36 +93,27 @@ end
 
 @register_serialization_type WeylGroup uses_id
 
+type_params(W::WeylGroup) = WeylGroup, root_system(W)
```
Stupid question maybe: It seems to me that the first return value of `type_params` is always the argument type. Is this a correct observation?
If yes, wouldn't it be possible to refactor that away into the calling function, so that `type_params` loses the first return value? Imo, this could make the code a bit more readable. If there is a reason for this, I am happy to learn about it

antonydellavecchia replied:
maybe there is a way to get a round it, but it was the only way I found I could get the new `save_type_params` to work. See main.jl

lgoettgens commented on src/Serialization/main.jl:
```diff
@@ -188,57 +171,133 @@ function save_typed_object(s::SerializerState, x::T, key::Symbol) where T
-    save_data_dict(s, :attrs) do
-      for attr in attrs_list(s, T)
-        has_attribute(obj, attr) && save_typed_object(s, get_attribute(obj, attr), attr)
+save_type_params(s::SerializerState, obj::Any) = save_type_params(s, type_params(obj)...)
```
```suggestion
save_type_params(s::SerializerState, obj::Any) = save_type_params(s, typeof(obj), type_params(obj)...)
```

I think this could do the trick

antonydellavecchia replied:
yes this is probably fine

antonydellavecchia replied:
actually, a little trickier than expected, there is an issue with the containers. 

I'll have to dig deeper, but I dont have time now

antonydellavecchia replied:
I've pushed a fix for this now

antonydellavecchia committed:
stuck on regression in polyhedral

antonydellavecchia committed:
simplify type_params

antonydellavecchia committed:
fields working

antonydellavecchia committed:
back to polyhedral

antonydellavecchia committed:
graded ring working

antonydellavecchia committed:
sets not passing yet

antonydellavecchia committed:
started to change docs a bit

antonydellavecchia committed:
fix some set tests

antonydellavecchia committed:
TypeParams Struct

antonydellavecchia committed:
progress on polyhedral objects

antonydellavecchia committed:
progress on some polyhedral object types

antonydellavecchia committed:
more progress

HechtiDerLachs committed:
Fix for type_params of dictionaries to avoid infinite recursion.

benlorenz committed:
serialization: flatten some polymake object dicts

antonydellavecchia committed:
fix dict type_params

antonydellavecchia committed:
qqbar fix

antonydellavecchia committed:
second line fix for QQbar Field

antonydellavecchia committed:
proper fix

antonydellavecchia committed:
fixes for container tests

antonydellavecchia committed:
some progress with TypeParams Struct

antonydellavecchia committed:
Algebras and Fields done

antonydellavecchia committed:
groups working again

antonydellavecchia committed:
lie theory + MPolyMap

antonydellavecchia committed:
bunch more types working

antonydellavecchia committed:
tests passing for all types in src (except upgrade test)

antonydellavecchia committed:
cleaning up polymake dict functions

lgoettgens committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
removed redundant function

antonydellavecchia committed:
Merge remote-tracking branch 'origin/adv/ipc-serialize' into adv/ipc-serialize

lgoettgens commented on src/Serialization/Groups.jl:
```diff
@@ -157,7 +139,7 @@ end
-  generators = load_object(s, Vector, (Vector{Int}, Int), :gens)
+  generators = load_object(s, Vector{Vector{Int}}, :gens)
   G = permutation_group(n, [perm(x) for x in generators])
   load_attrs(s, G)
```
```suggestion
```
this can go, right? And similarly the `save_attrs` 6 lines above

antonydellavecchia replied:
yes

antonydellavecchia committed:
Update src/Serialization/Groups.jl

Co-authored-by: Lars Göttgens <lars.goettgens@rwth-aachen.de>

lgoettgens committed:
Serialization refactor (Lie theory part) (#4563)

* Use `grading_group` for `MPolyDecRing`

* Fix some formatting

* Some progress

* Fix loading AbstractLieAlgebra

* LinearLieAlgebras working

* Modules working

* Attrs are handled centrally now

* Root system data working

* Module construction data works

* DirectSumLieAlgebras work

* oops

* Run formatter

* Reduce diff

lgoettgens commented:
In #4563, I encountered more instances of https://github.com/oscar-system/Oscar.jl/issues/3983. It would be great if you could have a look if the new loading code allows for an easy fix of this. (at some point in the refactor, not right now)

antonydellavecchia committed:
remove dead code

antonydellavecchia committed:
remove obsolete code see pr #4501

antonydellavecchia commented:
closes #4567

antonydellavecchia committed:
can almost load example qsm

antonydellavecchia committed:
qsm saving and loading refactor

antonydellavecchia committed:
starting on upgrade

lgoettgens committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
fix loading vectors

HereAround committed:
[FTheoryTools] Update serialization of Weierstrass models

HereAround committed:
[FTheoryTools] Update serialization of GlobalTateModel

antonydellavecchia committed:
allow dict overrides

antonydellavecchia committed:
upgrade fqPolyrepFields

antonydellavecchia committed:
some upgrades

lgoettgens commented on src/Serialization/Upgrades/1.3.0.jl:
```diff
@@ -1,16 +1,105 @@
+function upgrade_NamedTuple(d::Dict)
```
```suggestion
module SerializationUpgrade-1-3-0

function upgrade_NamedTuple(d::Dict)
```
It would be good to put all of these `upgrade_*` functions into their own shbmodule to not clutter the global namespace. And the next time someone needs to write an upgrade script, it could otherwise lead to name clashes.

antonydellavecchia replied:
This was only temporary because I thought that revise couldn't handle updates inside the function that got pushed into the function list.
Also, it was unsafe because I was evalutating strings from files

antonydellavecchia committed:
start on moving ftheory attributes to attribute framework

antonydellavecchia committed:
commented out setting some attributes

antonydellavecchia committed:
use tuples for resolutions

antonydellavecchia committed:
upgrade surfaces

antonydellavecchia committed:
upgrades for quad spaces

antonydellavecchia committed:
update enrique surfaces

antonydellavecchia committed:
Vector needs upgrade

antonydellavecchia committed:
issue with storing johnson solids after update

antonydellavecchia committed:
fix

antonydellavecchia committed:
comment out print statements

antonydellavecchia committed:
upgrade johnson solids

antonydellavecchia committed:
revert to using dev version on save

antonydellavecchia committed:
update catalan and archimedean solids

antonydellavecchia committed:
upgrades for all types in src

antonydellavecchia committed:
forgot to add file with ne extension

antonydellavecchia committed:
updates for loading IdealGens

antonydellavecchia committed:
improved Dict serialization coverage

antonydellavecchia committed:
catch missed upgrades

antonydellavecchia commented:
files of all types in src that have serialization tests.
[files_to_update.zip](https://github.com/user-attachments/files/18980076/files_to_update.zip)

antonydellavecchia committed:
Merge remote-tracking branch 'origin/master' into adv/ipc-serialize

antonydellavecchia committed:
Merge remote-tracking branch 'origin/master' into adv/ipc-serialize

antonydellavecchia committed:
fixes

lgoettgens committed:
Serialization Refactor (Lie Theory again) (#4649)

* Update LieAlgebra serialization again

* Progress towards Lie algebra module construction data

* Move direct summands to type params

* Simplify standard module

* Changes to construction data again

* handles TypeParams with tuples

* fix for non refs types in tuple

* Remaining fix

---------

Co-authored-by: Antony Della Vecchia <vecchia@math.tu-berlin.de>

antonydellavecchia committed:
update versions

antonydellavecchia committed:
upgrades ZZLatwithisom and some lie algebra upgrades

antonydellavecchia committed:
upgrades some lie algebra modules

antonydellavecchia committed:
fix vararg

antonydellavecchia committed:
only cast vectors when empty

antonydellavecchia committed:
small adjustment in uupgrades for ntvs without attrs

antonydellavecchia committed:
finish all upgrades for lie algebra modules

antonydellavecchia committed:
fix upgrade for ntv

antonydellavecchia committed:
Merge branch 'adv/ipc-serialize' into adv/refactor-ftheory-attributes

antonydellavecchia committed:
add all attributes

antonydellavecchia committed:
fixes regressions and can load qsm models

antonydellavecchia committed:
remove debugging print statements

antonydellavecchia committed:
fixes regression in polyhedron files

antonydellavecchia committed:
cleanup version forcing

antonydellavecchia committed:
only force dict type when empty

antonydellavecchia committed:
fix container tests

HereAround committed:
Update QSMDB

antonydellavecchia committed:
revert change in polynomial example

antonydellavecchia committed:
fix the ipc init

antonydellavecchia committed:
Merge remote-tracking branch 'origin/adv/ipc-serialize' into adv/ipc-serialize

antonydellavecchia committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
skip literature model tests for now

antonydellavecchia committed:
remove print on polyhedralobject upgrade

antonydellavecchia committed:
started on artifacts

antonydellavecchia committed:
started on artifact

antonydellavecchia committed:
move things into literature tests to avoid any kind of loading

lgoettgens commented:
Would it be possible to split this PR into two separate ones before merging? The github UI doesn't quite like >300 changed files, thus it is not possible to look at the changes that this introduces.
If I understand it correctly, it should be possible to split into the following two parts:
1. All changes to the serialization, upgrade scripts etc.
2. Upgrading everything in `data/` and the artifacts

As the upgrade scripts are contained in 1, the testsuite should even succeed in-between the two PRs. And then one would only need to look at 1 to see the "real" changes. But all of the many completely changed files that github trips over would be part of 2.

I imagine this not being too difficult, but I could be wrong.

antonydellavecchia committed:
upgrade tests for all files in version 1.3.0

antonydellavecchia committed:
revert files

antonydellavecchia committed:
add lazy artifacts to deps

antonydellavecchia committed:
skip tate model serialization tests

antonydellavecchia committed:
skip another ftheory test

antonydellavecchia committed:
revert lp to old format

antonydellavecchia committed:
fixes for dicts of diff value types

HereAround committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
fix for int keys

antonydellavecchia committed:
typo

antonydellavecchia committed:
catch empty and not equl params

antonydellavecchia committed:
fix dicts

antonydellavecchia committed:
properly upgrades polyhedral objects over number field

antonydellavecchia committed:
fixes loading tests

antonydellavecchia committed:
update artifact

antonydellavecchia committed:
fix dir_name

antonydellavecchia committed:
update no experimental tests

antonydellavecchia committed:
FTheory Tools attributes  (#4620)

* More attributes

* Take care of resolutions

* Complete list of attributes

* Introduce some dummy attributes

* Update QSMDB

* Update QSMDB

* Fix loading of new QSM models

* Remove obsolete code (was used to load old version of QSM models)

* Fix tests

* Fix more tests

* Fix more tests

* Further upgrade to new attribute framework

* Move conversion of inter_dict to other places

* Mirror changes for hypersurface models also for Weierstrass and Tate models

* White space

* Try to fix upgrading tate models

* Attempt to fix serialization of GlobalTateModels

* Present weighted resolutions in slightly different format

* More improvements to the serialization of Tate models

* Check if Tate polynomial is known

* adjust dict serialization

* fixes for dicts of diff value types

* Attempt to fix upgrade script for GlobalTateModels (but failed of course...)

* progress

* More attempts to get upgrade script working

* fix tate model upgrade

* Attempt to fix upgrade script for GlobalTateModels (but failed of course...)

* fix tate model upgrade

* More attempts to get upgrade script working

* fix for dict{Int, U}

* Remove outcommented stuff

* Improve hypersurface model serialization

* Extend upgrade script to Weierstrass and hypersurface models

* Improve weierstrass model serialization

* Minor fix in loading/saving of Weierstrass models

* upgrades everything except polyhedral objects over field extensions

* Update printing to avoid changed output on different OS

* Reflect internal change to weighted resolutions in CI test

* Remove additional "end"

* Fix change of weighted_resolution in doctest

* Bugfix

* Fix doctests

* fix incorrectly resolved conflict

* fix upgrade regression

* Fix for CI test

---------

Co-authored-by: Martin Bies <MBies87@googlemail.com>
Co-authored-by: Martin Bies <HereAround@users.noreply.github.com>

benlorenz commented on src/Combinatorics/SimplicialComplexes.jl:
```diff
@@ -637,7 +637,7 @@ function deletion(K::SimplicialComplex, face::Union{<:AbstractSet{Int},<:Abstrac
 
 @doc raw"""
-    automorphism_group(K::SimplicialComplex; action=:on_vertices)
+    automorphism_group(K::SimplicialComplex)
```
```suggestion
    automorphism_group(K::SimplicialComplex; action=:on_vertices)
```

benlorenz commented on src/Combinatorics/SimplicialComplexes.jl:
```diff
@@ -670,7 +670,7 @@ function automorphism_group(K::SimplicialComplex; action=:on_vertices)
 
 @doc raw"""
-    on_simplicial_complex(K::SimplicialComplex, g::PermGroupElem)
+    on_simplicial_complex(K::SimplicialComplex, g::PermGroupElem;)
```
```suggestion
    on_simplicial_complex(K::SimplicialComplex, g::PermGroupElem)
```

antonydellavecchia commented on src/PolyhedralGeometry/helpers.jl:
```diff
@@ -317,7 +317,11 @@ function (NF::Hecke.EmbeddedNumField)(x::Polymake.QuadraticExtension{Polymake.Ra
+  #println(Oscar.global_serializer_state.obj_to_id[F])
+  #println(Oscar.global_serializer_state.obj_to_id[parent(Polymake.unwrap(x))])
+  F(Polymake.unwrap(x))
+end
```
```suggestion
(F::Field)(x::Polymake.OscarNumber) = F(Polymake.unwrap(x))
```

benlorenz replied:
```suggestion
(F::Field)(x::Polymake.OscarNumber) = F(Polymake.unwrap(x))
```

benlorenz commented on src/Serialization/Upgrades/1.2.2.jl:
please change this file back to 1.3.0 (filename + content) to make it consistent with the 1.3.0 release

benlorenz commented on src/Serialization/Upgrades/main.jl:
```diff
@@ -141,7 +141,8 @@ include("0.13.0.jl")
 include("1.1.0.jl")
 include("1.2.0.jl")
-include("1.3.0.jl")
+include("1.2.2.jl")
```
```suggestion
include("1.3.0.jl")
```

antonydellavecchia committed:
Apply suggestions from code review

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

antonydellavecchia committed:
renaming file

antonydellavecchia committed:
revert IPC test

lgoettgens commented:
The macos failures are probably due to https://github.com/oscar-system/Oscar.jl/issues/4714

antonydellavecchia committed:
remove debugging artifact

antonydellavecchia committed:
one more debuggin artifact

fingolfin reviewed (approved):
Fine by me.

lgoettgens commented:
@antonydellavecchia could you please come up with approx. 2 sentences about this PR that we can put into the v1.4 release notes?

antonydellavecchia commented:
Cleaner handling of type parameter serialization. This update forces entries of container types to share the same output of Oscar.type_params when serializing. Deserialization speed improvements.
